server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java

```diff
@@ -394,21 +397,288 @@ public void testTextSubFields() throws IOException {
+            .value("string_value")
+            .startObject()
+                .field("a", "b")
+            .endObject()
+            .startArray()
+                .value(1).value(2).value(3)
+            .endArray()
+            .endArray()
+        .endObject();
+
+        Map<String, DocumentField> fields = fetchFields(mapperService, source, "*", false);
+        // without unmapped fields this should return nothing
+        assertThat(fields.size(), equalTo(0));
+
+        fields = fetchFields(mapperService, source, "*", true);
+        assertThat(fields.size(), equalTo(2));
+        assertThat(fields.keySet(), containsInAnyOrder("obj", "obj.a"));
+
+        List<Object> obj = fields.get("obj").getValues();
+        assertEquals(2, obj.size());
+        assertThat(obj.get(0), instanceOf(String.class));
+        assertEquals("string_value", obj.get(0).toString());
+        assertThat(obj.get(1), instanceOf(List.class));
+        assertEquals(3, ((List<?>) obj.get(1)).size());
+        assertEquals("[1, 2, 3]", obj.get(1).toString());
+
+        List<Object> innerObj = fields.get("obj.a").getValues();
+        assertEquals(1, innerObj.size());
+        assertEquals("b", fields.get("obj.a").getValue());
+    }
+
+    /**
+     * If a mapped field for some reason contains a "_source" value that is not returned by the
+     * mapped retrieval mechanism (e.g. because its malformed), we don't want to fetch it from _source.
+     */
+    public void testMappedFieldNotOverwritten() throws IOException {
+        XContentBuilder mapping = XContentFactory.jsonBuilder().startObject()
+            .startObject("properties")
+                .startObject("f1")
+                    .field("type", "integer")
+                    .field("ignore_malformed", "true")
+                .endObject()
+            .endObject()
+        .endObject();
+
+        IndexService indexService = createIndex("index", Settings.EMPTY, mapping);
+        MapperService mapperService = indexService.mapperService();
 
+        XContentBuilder source = XContentFactory.jsonBuilder().startObject()
+            .field("f1", "malformed")
+            .endObject();
+
+        // this should not return a field bc. f1 is in the ignored fields
+        Map<String, DocumentField> fields = fetchFields(
+            mapperService,
+            source,
+            List.of(new FieldAndFormat("*", null, Optional.of(true))),
+            Set.of("f1")
+        );
+        assertThat(fields.size(), equalTo(0));
+
+        // and this should neither
+        fields = fetchFields(mapperService, source, List.of(new FieldAndFormat("*", null, Optional.of(true))), Set.of("f1"));
+        assertThat(fields.size(), equalTo(0));
+
+        fields = fetchFields(mapperService, source, List.of(new FieldAndFormat("f1", null, Optional.of(true))), Set.of("f1"));
+        assertThat(fields.size(), equalTo(0));
+    }
+
+    public void testUnmappedFieldsWildcard() throws IOException {
```

Comment:
The previous test case 'testMappedFieldNotOverwritten' tests the special case where we don't want to return a mapped field that is not present in the document but in `_source` (e.g. because of ignore_malformed). Here I wanted to test various field pattern behaviours for object structures in an isolated test. So I think its not ideal to merge them.